### PR TITLE
Dynamic power plant models, examples, tests, and documentation

### PR DESCRIPTION
**Pull request by MAZamarripa**
## Fixes
update power plant models, add dynamic model of FWH, add dynamic power plant example, documentation, and tests

## Summary/Motivation:
complete dynamic power plant library and examples

## Changes proposed in this PR:
- add dynamic power plant example, documentation, and tests
- update power plant models (downcomer initialize, waterwall, turbine) 
- add dynamic model of FWH, 
- update boiler fireside model (one parameter)
- update flue gas properties (one parameter only)

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: idaes/power_generation/flowsheets/subcritical_power_plant/subcritical_power_plant.py
Comment (eslickj) [inline idaes/power_generation/flowsheets/subcritical_power_plant/subcritical_power_plant.py:46]: This is causing the failing test.  I thought we were going to remove this.  If not change this line to:

```python
    if kwargs.get("tee", False):
```

Comment (eslickj) [reply idaes/power_generation/flowsheets/subcritical_power_plant/subcritical_power_plant.py]: Let's drop this weirdness once PR #180 gets merged.

File: docs/technical_specs/model_libraries/power_generation/flowsheets/subcritical_power_plant.rst
Comment (eslickj) [inline docs/technical_specs/model_libraries/power_generation/flowsheets/subcritical_power_plant.rst:122]: You can link the unit library name to the model docs.

File: docs/technical_specs/model_libraries/power_generation/flowsheets/subcritical_power_plant.rst
Comment (eslickj) [inline docs/technical_specs/model_libraries/power_generation/flowsheets/subcritical_power_plant.rst:106]: This is a little painful, but you may want to actually break the lines of text rather than using word wrap.  It would make it easier to make review comments on specific lines.

File: docs/technical_specs/model_libraries/power_generation/flowsheets/subcritical_power_plant.rst
Comment (eslickj) [reply docs/technical_specs/model_libraries/power_generation/flowsheets/subcritical_power_plant.rst:37]: found a linking example as I was reviewing 

```rst
:ref:`IAWPS95 <technical_specs/model_libraries/generic/property_models/water:Water/Steam - IAPWS95>`
```

**Review by eslickj — COMMENTED** [review]
Just got through a chunk of docs.  Will continue review soon.

File: docs/technical_specs/model_libraries/power_generation/flowsheets/subcritical_power_plant.rst
Comment (Morgan88888888) [reply docs/technical_specs/model_libraries/power_generation/flowsheets/subcritical_power_plant.rst:12]: I agree. I think we could point to the PFD template, like adding file name "plant_pfd.svg". It's also good to paste the png version I generated.

Comment (Daison2102) [reply docs/technical_specs/model_libraries/power_generation/flowsheets/subcritical_power_plant.rst:12]: Maybe just need to mention that PFD is in Figure 1

File: docs/technical_specs/model_libraries/power_generation/flowsheets/subcritical_power_plant.rst
Comment (Daison2102) [inline docs/technical_specs/model_libraries/power_generation/flowsheets/subcritical_power_plant.rst:100]: You should change the word Figures to Figure 1

File: docs/technical_specs/model_libraries/power_generation/flowsheets/subcritical_power_plant.rst
Comment (Daison2102) [inline docs/technical_specs/model_libraries/power_generation/flowsheets/subcritical_power_plant.rst:188]: Captions and numbers for Figures 2 and 3 are missing

File: idaes/power_generation/flowsheets/subcritical_power_plant/steam_cycle_flowsheet.py
Comment (eslickj) [reply idaes/power_generation/flowsheets/subcritical_power_plant/steam_cycle_flowsheet.py:1710]: Of maybe generate text file if they are important results?

File: idaes/power_generation/flowsheets/subcritical_power_plant/subcritical_boiler_flowsheet.py
Comment (eslickj) [inline idaes/power_generation/flowsheets/subcritical_power_plant/subcritical_boiler_flowsheet.py:1096]: Again maybe log if this isn't the main results or if they are super important results, maybe send to a text file.

File: idaes/power_generation/flowsheets/subcritical_power_plant/subcritical_boiler_flowsheet.py
Comment (eslickj) [inline idaes/power_generation/flowsheets/subcritical_power_plant/subcritical_boiler_flowsheet.py:1207]: Some of these prints should definitely be log.

**Review by eslickj — COMMENTED** [review]
The code looks pretty good.  Made some comments.  Still need to carefully read docs.

File: idaes/power_generation/unit_models/tests/test_feedwater_heater_dynamic.py
Comment (eslickj) [inline idaes/power_generation/unit_models/tests/test_feedwater_heater_dynamic.py:31]: We may want to break the tests up so that everything up to the initialization and solve is a unit test the initialization and solve are integration tests.  That will help catch a lot of things before running the integration tests.  I think if you write a test class you can just break up what you have into a sequential set of tests @andrewlee94 may have a recommendation to a good example to follow.

Comment (andrewlee94) [reply idaes/power_generation/unit_models/tests/test_feedwater_heater_dynamic.py:31]: The CSTR is always my go-to example.

**Comment by eslickj** [comment]
I think this is pretty close.  It looks like you are still monkey patching ipopt.  We should take that out now.  If it causes convergence issues, we can accomplish the same with PR #180.  Other than that the only thing may be minor style things.  I'll get it reviewed again today.

File: idaes/power_generation/control/pid_controller.py
Comment (andrewlee94) [inline idaes/power_generation/control/pid_controller.py:14]: A general question ,but how does this differ from the PID controller in the core library?

File: idaes/power_generation/control/pid_controller.py
Comment (andrewlee94) [inline idaes/power_generation/control/pid_controller.py:68]: Very minor, but would it have been clearer to do:

```
if not self.config.dynamic:
    return
```

Also, should we log a message to say we did mothing (info_high or debug level)?

File: idaes/power_generation/control/tests/test_pid_pump.py
Comment (andrewlee94) [inline idaes/power_generation/control/tests/test_pid_pump.py:54]: This is a really weird way to write tests.

For one, I would suggest two separate tests for the steady-state and dynamic cases to better identify where the problem is if something breaks.

Secondly, there really should be unit tests in here as well to make sure the models build correctly and look right before you try solving them.

**Review by andrewlee94 — COMMENTED** [review]
I did not look through with my normal level of detail due to lack of time, but most things looked OK at first glance. However, I note that there are not many tests, especially for the flowsheets and the tests for the PID unit do not include any unit tests.

File: idaes/power_generation/control/tests/test_pid_pump.py
Comment (MAZamarripa) [reply idaes/power_generation/control/tests/test_pid_pump.py:54]: we followed the example of the generic pid controller
this is more like a flowsheet example any test ends being an integration test

File: idaes/power_generation/control/pid_controller.py
Comment (eslickj) [reply idaes/power_generation/control/pid_controller.py:14]: This one has a variable for the integral error term.  The core library now uses the velocity form to avoid the need for it, but I'm not sure why one may be better than the other and when to prefer one over the other.  This also uses different functions to limit the controller output between bounds.  Also this supports P and PD.

Eventually when I have time, I'll merge those things in as options to the core controller.  In the interest of time, we are going to hold off on the merger to a single PID class.

**Review by eslickj — APPROVED** [review]

**Review by Daison2102 — APPROVED** [review]

File: idaes/power_generation/control/tests/test_pid_pump.py
Comment (andrewlee94) [reply idaes/power_generation/control/tests/test_pid_pump.py:54]: My point is that there really should be a separate set of unit tests of the controller - these do not need to be big for complicated. Jut build a flowsheet, put two `Vars` on it (the measured and controlled variables), add an instance of the controller and check that it has the right set of constraints. do this for dynamic and steady-state and you are done.

**Comment by andrewlee94** [comment]
@MAZamarripa There appears to be a problem in your tests, as all the integration tests are failing - one of the values in the dynamic feedwater tests appears to be wrong.

**Review by andrewlee94 — DISMISSED** [review]
Rubber stamps, although I hope to see more tests in the future.

**Review by eslickj — APPROVED** [review]

**Review by eslickj — APPROVED** [review]

**Review by andrewlee94 — APPROVED** [review]
Looks good enough for sheep stations.

**Pull request merged**
